// beeai-framework/dist/backend/chat.js
import { Serializable } from '../internals/serializable.js';
import { shallowCopy } from '../serializer/utils.js';
import { customMerge } from '../internals/helpers/object.js';
import { takeBigger } from '../internals/helpers/number.js';
import { RunContext } from '../context.js';
import { INSTRUMENTATION_ENABLED } from '../instrumentation/config.js';
import { createTelemetryMiddleware } from '../instrumentation/create-telemetry-middleware.js';
import { doNothing, isFunction } from 'remeda';
import { ObjectHashKeyFn } from '../cache/decoratorCache.js';
import { Task } from 'promise-based-task';
import { NullCache } from '../cache/nullCache.js';
import { parseModel, loadModel } from './utils.js';
import { SystemMessage, AssistantMessage, UserMessage } from './message.js';
import { ChatModelError } from './errors.js';
import { z } from 'zod';
import { toJsonSchema, parseBrokenJson, createSchemaValidator } from '../internals/helpers/schema.js';
import { Retryable } from '../internals/helpers/retryable.js';
import { PromptTemplate } from '../template.js';
import { toAsyncGenerator } from '../internals/helpers/promise.js';
import { Serializer } from '../serializer/serializer.js';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class ChatModel extends Serializable {
  static {
    __name(this, "ChatModel");
  }
  cache = new NullCache();
  parameters = {};
  create(input) {
    input = shallowCopy(input);
    return RunContext.enter(this, {
      params: [
        input
      ],
      signal: input?.abortSignal
    }, async (run) => {
      const cacheEntry = await this.createCacheAccessor(input);
      try {
        await run.emitter.emit("start", {
          input
        });
        const chunks = [];
        const generator = cacheEntry.value ?? (input.stream ? this._createStream(input, run) : toAsyncGenerator(this._create(input, run)));
        const controller = new AbortController();
        for await (const value of generator) {
          chunks.push(value);
          await run.emitter.emit("newToken", {
            value,
            callbacks: {
              abort: /* @__PURE__ */ __name(() => controller.abort(), "abort")
            }
          });
          if (controller.signal.aborted) {
            break;
          }
        }
        cacheEntry.resolve(chunks);
        const result = ChatModelOutput.fromChunks(chunks);
        await run.emitter.emit("success", {
          value: result
        });
        return result;
      } catch (error) {
        await run.emitter.emit("error", {
          input,
          error
        });
        await cacheEntry.reject(error);
        if (error instanceof ChatModelError) {
          throw error;
        } else {
          throw new ChatModelError(`The Chat Model has encountered an error.`, [
            error
          ]);
        }
      } finally {
        await run.emitter.emit("finish", null);
      }
    }).middleware(INSTRUMENTATION_ENABLED ? createTelemetryMiddleware() : doNothing());
  }
  createStructure(input) {
    return RunContext.enter(this, {
      params: [
        input
      ],
      signal: input?.abortSignal
    }, async (run) => {
      return await this._createStructure(input, run);
    });
  }
  config({ cache, parameters }) {
    if (cache) {
      this.cache = isFunction(cache) ? cache(this.cache) : cache;
    }
    if (parameters) {
      this.parameters = isFunction(parameters) ? parameters(this.parameters) : parameters;
    }
  }
  static async fromName(name, options) {
    const { providerId, modelId } = parseModel(name);
    const Target = await loadModel(providerId, "chat");
    return new Target(modelId || void 0, options);
  }
  async _createStructure(input, run) {
    const { schema, ...options } = input;
    const jsonSchema = toJsonSchema(schema);
    const systemTemplate = new PromptTemplate({
      schema: z.object({
        schema: z.string().min(1)
      }),
      template: `You are a helpful assistant that generates only valid JSON adhering to the following JSON Schema.

\`\`\`
{{schema}}
\`\`\`

IMPORTANT: You MUST answer with a JSON object that matches the JSON schema above.`
    });
    const messages = [
      new SystemMessage(systemTemplate.render({
        schema: JSON.stringify(jsonSchema, null, 2)
      })),
      ...input.messages
    ];
    const errorTemplate = new PromptTemplate({
      schema: z.object({
        errors: z.string(),
        expected: z.string(),
        received: z.string()
      }),
      template: `Generated object does not match the expected JSON schema!

Validation Errors: {{errors}}`
    });
    return new Retryable({
      executor: /* @__PURE__ */ __name(async () => {
        const response = await this._create({
          ...options,
          messages,
          responseFormat: {
            type: "object-json"
          }
        }, run);
        const textResponse = response.getTextContent();
        const object = parseBrokenJson(textResponse, {
          pair: [
            "{",
            "}"
          ]
        });
        const validator = createSchemaValidator(schema);
        const success = validator(object);
        if (!success) {
          const context = {
            expected: JSON.stringify(jsonSchema),
            received: textResponse,
            errors: JSON.stringify(validator.errors ?? [])
          };
          messages.push(new UserMessage(errorTemplate.render(context)));
          throw new ChatModelError(`LLM did not produce a valid output.`, [], {
            context
          });
        }
        return {
          object
        };
      }, "executor"),
      config: {
        signal: run.signal,
        maxRetries: input?.maxRetries || 1
      }
    }).get();
  }
  createSnapshot() {
    return {
      cache: this.cache,
      emitter: this.emitter,
      parameters: shallowCopy(this.parameters)
    };
  }
  destroy() {
    this.emitter.destroy();
  }
  async createCacheAccessor({ abortSignal: _, messages, tools = [], ...input }) {
    const key = ObjectHashKeyFn({
      ...input,
      messages: await Serializer.serialize(messages.map((msg) => msg.toPlain())),
      tools: await Serializer.serialize(tools)
    });
    const value = await this.cache.get(key);
    const isNew = value === void 0;
    let task = null;
    if (isNew) {
      task = new Task();
      await this.cache.set(key, task);
    }
    return {
      key,
      value,
      resolve: /* @__PURE__ */ __name((value2) => {
        task?.resolve?.(value2);
      }, "resolve"),
      reject: /* @__PURE__ */ __name(async (error) => {
        task?.reject?.(error);
        if (isNew) {
          await this.cache.delete(key);
        }
      }, "reject")
    };
  }
}
class ChatModelOutput extends Serializable {
  static {
    __name(this, "ChatModelOutput");
  }
  messages;
  usage;
  finishReason;
  constructor(messages, usage, finishReason) {
    super(), this.messages = messages, this.usage = usage, this.finishReason = finishReason;
  }
  static fromChunks(chunks) {
    const final = new ChatModelOutput([]);
    chunks.forEach((cur) => final.merge(cur));
    return final;
  }
  merge(other) {
    this.messages.push(...other.messages);
    this.finishReason = other.finishReason;
    if (this.usage && other.usage) {
      this.usage = customMerge([
        this.usage,
        other.usage
      ], {
        totalTokens: takeBigger,
        promptTokens: takeBigger,
        completionTokens: takeBigger
      });
    } else if (other.usage) {
      this.usage = shallowCopy(other.usage);
    }
  }
  getToolCalls() {
    return this.messages.filter((r) => r instanceof AssistantMessage).flatMap((r) => r.getToolCalls()).filter(Boolean);
  }
  getTextMessages() {
    return this.messages.filter((r) => r instanceof AssistantMessage).filter((r) => r.text);
  }
  getTextContent() {
    return this.messages.filter((r) => r instanceof AssistantMessage).flatMap((r) => r.text).filter(Boolean).join("");
  }
  toString() {
    return this.getTextContent();
  }
  createSnapshot() {
    return {
      messages: shallowCopy(this.messages),
      usage: shallowCopy(this.usage),
      finishReason: this.finishReason
    };
  }
  loadSnapshot(snapshot) {
    Object.assign(this, snapshot);
  }
}

export { ChatModel, ChatModelOutput };
